File: python/src/entities/BinarySearchTree/TreeNode.py
from typing import List

class TreeNode:
    # forward reference class TreeNode as it has not been defined yet
    parent: 'TreeNode' = None
    # Left child
    left: 'TreeNode' = None
    # Right child
    right: 'TreeNode' = None
    # The actual value stored in the node
    val: int = None
    # The depth of the node in the tree. Root has depth of 0.
    depth: int = 0
    
    
    def __init__(self, node_numbers: List[int], parent: 'TreeNode' = None) -> 'TreeNode':
        """Constructs a TreeNode from a list of integers.

        Args:
            node_numbers (List[int]): Sorted, unique list of integers to construct a `TreeNode` from.
            parent (TreeNode, optional): Parent node to set on this `TreeNode`. Defaults to None.

        Returns:
            TreeNode: The root node of the constructed tree.
        """
        
        # No integers are left. This means the parent node is a leaf node.
        if len(node_numbers) == 0:
            raise ValueError("Cannot construct a tree from an empty list of integers.")
        
        # Get the middle index of the list of integers
        root_index = len(node_numbers) // 2
        root_node_number = node_numbers[root_index]
        
        # Split the list of integers into two halves
        left_subtree = node_numbers[:root_index]
        right_subtree = node_numbers[root_index + 1:]
        
        self.val = root_node_number
        self.parent = parent
        self.depth = parent.depth + 1 if parent is not None else 0
        
        # Recursively construct the left and right subtrees
        self.left = TreeNode(left_subtree, self) if len(left_subtree) > 0 else None
        self.right = TreeNode(right_subtree, self) if len(right_subtree) > 0 else None
        
    @staticmethod
    def equals(a: 'TreeNode', b: 'TreeNode') -> bool:
        """Returns whether two treenodes are equal

        Returns:
            bool: Whether these two treenodes are exactly equal
        """
        
        # Since we assume that all node values are unique, all we need to know is whether the values are equal
        return a.val == b.val
    
    def preprocess_val_for_report(self, val: int) -> str:
        """Preprocesses a value for printing in the report.

        Args:
            val (int): The value to preprocess.

        Returns:
            str: The preprocessed value.
        """
        retVal = str(val)
        
        # Surround negative values in parentheses
        if val < 0:
            retVal = f"({retVal})"
        
        return retVal
    def print_report(self, depth = 0):
        """Prints the node and its descendants in pre-order traversal.

        Args:
            depth (int, optional): Depth of the printed report. 
                The first level of the report will include a short description of report format.
                Defaults to 0.
        """
        if (depth == 0):
            print("Printing tree of cardinality " + str(len(self.descendants()) + 1) + " in pre-order traversal.")
            print("At each level, the left node is printed first, then the right node.")
            
        # Add some lines to the front of the string to indicate the depth of the node
        prefix_tree_lines = '-' * self.depth * 2
        
        # printed string starts off with the prefix tree lines and the value of the node
        print_str = F"{prefix_tree_lines}{self.preprocess_val_for_report(self.val)}"
        
        # Add some subtext to the printed string to indicate if the node is the root or a leaf
        if (self.is_root()):
            print_str += "\t(ROOT)"
            
        if not self.is_root():
            if TreeNode.equals(self, self.parent.left):
                print_str += "\t(LEFT CHILD)"
            elif TreeNode.equals(self, self.parent.right):
                print_str += "\t(RIGHT CHILD)"
            if len(self.parent.children()) == 1:
                print_str += "\t(ONLY CHILD)"
                
        if self.is_leaf():
            print_str += "\t(LEAF)"
            
        print(print_str)
        
        # Print children recursively
        for child in self.children():
            child.print_report(depth + 1)

    def is_root(self) -> bool:
        """Returns whether this node is the root of the tree.

        Returns:
            bool: Whether this node is a tree root (has no parent)
        """
        return self.parent is None
    
    def is_leaf(self) -> bool:
        """Returns whether this node is a leaf node (has no children).

        Returns:
            bool: Whether this node is a leaf node
        """
        return len(self.children()) == 0

    def children(self) -> List['TreeNode']:
        """Returns a list of the children of this node.

        Returns:
            _type_: All children of this node.
        """
        return [node for node in [self.left, self.right] if node is not None]
    
    def ancestors(self) -> List['TreeNode']:
        """Returns a list of all ancestors (parent, grandparent, etc.) of this node.

        Returns:
            _type_: All ancestors of this node.
        """
        retVal = []
        if self.is_root():
            return retVal
        
        retVal.append(self.parent)
        retVal.extend(self.parent.ancestors())
            
        return retVal

    def descendants(self) -> List['TreeNode']:
        """Returns a list of all descendants (children, children's children, etc.) of this node.

        Returns:
            _type_: All descendants of this node.
        """
        retVal = []
        for child in self.children():
            retVal.append(child)
            retVal.extend(child.descendants())

        return retVal
    
    def search(self, val: int) -> 'TreeNode':
        """Searches the tree for a node with the given value.

        Args:
            val (int): Value to search for.

        Returns:
            TreeNode: Node with `val`=`val`, if found. Else, `None`.
        """
        
        assert self is not None, "Cannot search a null tree."
        
        # Base case 1: This node has the value we're looking for! Return it.
        if self.val == val:
            return self
        
        # Base case 2: This node is a leaf node. This branch of the search is over, return None.
        if self.is_leaf():
            return None
        
        # Decide on left or right subtree based on numeric comparison
        subtree = self.left if val < self.val else self.right
        
        # If the subtree is None and we've gotten this far, the node only has one child, but that child mathematically cannot have the value we're looking for.
        # So the value is not in the tree.
        if subtree is None:
            return None
        
        # Recursive case: Search the appropriate subtree
        return subtree.search(val)

File: typescript/src/entities/TreeNode.ts
import { TreeNode as PrimeTreeNode } from 'primeng/api';

export const DEFAULT_PRIME_TREE_NODE: PrimeTreeNode<TreeNode> = {
    label: '',
    data: undefined,
    children: []
}

/**
 * Represents a node in a binary tree.
 */
export class TreeNode {
    /** Node's numeric content */
    val: number;

    parent?: TreeNode;

    /** Left subtree */
    left?: TreeNode;

    /** Right subtree */
    right?: TreeNode;

    /** Depth of the node in the tree. Root has depth of 0. */
    depth!: number;

    /** PrimeNG-compatibile treenode for org-chart style displaying */
    primeNode!: PrimeTreeNode<TreeNode>;

    constructor(nodeNumbers: number[], parent?: TreeNode) {
        if (!nodeNumbers.length) throw new Error("Cannot construct a tree node from an empty list.");

        const rootIndex = Math.floor(nodeNumbers.length / 2);
        const rootNumber = nodeNumbers[rootIndex];

        const leftSubTree = nodeNumbers.slice(0, rootIndex);
        const rightSubTree = nodeNumbers.slice(rootIndex + 1);

        this.val = rootNumber;
        this.parent = parent;
        this.depth = parent ? parent.depth + 1 : 0;
        this.primeNode = DEFAULT_PRIME_TREE_NODE;


        if (leftSubTree.length) this.left = new TreeNode(leftSubTree, this);
        if (rightSubTree.length) this.right = new TreeNode(rightSubTree, this);

        this.populatePrimeNode();
    }

    /**
     * Returns whether two tree nodes are equal.
     * @param a First tree node.
     * @param b Second tree node.
     * @returns Whether these two tree nodes are exactly equal.
     */
    static equals(a: TreeNode | null, b: TreeNode | null): boolean {
        // We can assume that tree nodes are unique, so we can compare them by value
        return a?.val === b?.val;
    }

    /**
     * Returns whether this node is the root of the tree.
     * @returns Whether this node is a tree root (has no parent).
     */
    isRoot(): boolean {
        return !!!this.parent;
    }

    /**
     * Returns whether this node is a leaf node (has no children).
     * @returns Whether this node is a leaf node.
     */
    isLeaf(): boolean {
        return !this.children().length;
    }

    /**
     * Returns a list of the children of this node.
     * @returns All children of this node.
     */
    children(): TreeNode[] {
        const allChildren = [this.left, this.right];
        const truthyChildren: TreeNode[] = allChildren.filter(node => !!node) as TreeNode[];
        return truthyChildren
    }

    /**
     * Returns a list of all ancestors (parent, grandparent, etc.) of this node.
     * @returns All ancestors of this node.
     */
    ancestors(): TreeNode[] {
        if (!this.parent) return [];

        return [this.parent, ...this.parent.ancestors()];
    }

    /**
     * Returns a list of all descendants (children, children's children, etc.) of this node.
     * @returns All descendants of this node.
     */
    descendants(): TreeNode[] {
        return this.children().reduce(
            (acc, child) => [...acc, child, ...child.descendants()], 
            new Array<TreeNode>()
        );
    }

    /**
     * Searches the tree for a node with the given value.
     * @param val Value to search for.
     * @returns Node with `val`=`val`, if found. Else, `null`.
     */
    search(val: number): TreeNode | null {
        if (!this) throw new Error("Cannot search a null tree.");
        
        // Base case 1: This node has the value we're looking for! Return it.
        if (this.val === val) return this;

        // Base case 2: This node is a leaf node. This branch of the search is over, return None.
        if (this.isLeaf()) return null;

        // Decide on left or right subtree based on numeric comparison
        const subtree = val < this.val ? this.left : this.right;
        
        if (!subtree) return null;
        
        // Recursive case: Search the appropriate subtree
        return subtree.search(val);
    }

    populatePrimeNode(): void {
        this.primeNode = this.toPrimeTreeNode() || DEFAULT_PRIME_TREE_NODE;
    }

    /**
     * Converts the tree to a PrimeNG-compatible treenode.
     * @returns PrimeNG org chart tree node
     */
    toPrimeTreeNode(): PrimeTreeNode {
        return {
            label: this.val.toString(),
            data: this,
            expanded: true,
            styleClass: this.getDOMClass(),
            children: this.children().map(child => child.toPrimeTreeNode())
        }
    }

    getDOMClass(): string {
        return `orgchart-item-${this.val.toString()}`   
    }

    getDOMNode(): HTMLDivElement | null {
        const matchingDOMNodes = document.getElementsByClassName(this.getDOMClass());
        if (!matchingDOMNodes?.length) return null;
        return matchingDOMNodes[0] as HTMLDivElement;
    }
}
